lib/styleToFilters.js

```diff
@@ -42,13 +45,98 @@ function styleToFilters(style) {
           layers[layerName].minzoom = style.layers[i].minzoom || 0;
           layers[layerName].maxzoom = style.layers[i].maxzoom || 22;
         }
+
+        // Collect the used property
+        layers[layerName].properties = layers[layerName].properties || [];
+        ['paint', 'layout'].forEach(item => {
+          let itemObject = style.layers[i][item];
+          itemObject && getPropertyLoop(itemObject, layers[layerName].properties);
+        });
       }
     }
   }
 
+  // remove duplicate propertys and fix choose all the propertys in layers[i].properties
+  Object.keys(layers).forEach(layerId => {
+    let properties = layers[layerId].properties;
+    if (properties.indexOf(true) !== -1) {
+      layers[layerId].properties = true;
+    } else {
+      let unique = {};
+      properties.forEach(function(i) {
+        if (!unique[i]) {
+          unique[i] = true;
+        }
+      });
+      layers[layerId].properties = Object.keys(unique);
+    }
+  })
+
   return layers;
 }
 
+
+function getPropertyLoop(propertyObj, properties) {
+  Object.keys(propertyObj).forEach(key => {
+    let value = propertyObj[key];
+    // TODO we still have outher situation:
+    // - special properties: `mapbox_clip_start`, `mapbox_clip_end`
+    if (typeof value === 'string') {
+      // if the value is string try to get property name from `xx{PropertyName}xx` like.
+      // the /{[^}]+}/ig return all the value like {xxx}
+      // eg 'a{hello}badfa' => ['{hello}']
+      // eg 'a{hello}ba{world}dfa' => ['{hello}','{world}']
+      let preProperties = value.match(/{[^}]+}/ig);
+      preProperties && preProperties.forEach(item => {
+        properties.push(item.slice(1, -1));
+      });
+    } else if (typeof value === 'object' && typeof value.property === 'string') {
+      // - legacy functions with `property`
+      properties.push(value.property);
+    } else {
+      // test isExpression from sytleSpec
+      if (styleSpec.expression.isExpression(value)) {
+        // TODO: now we implement this by ourself in vtshavem, we need to talk with ‘style spec’ member to see if there have a official method to get used property, to make this can be synchronized with the expression update.
+        getPropertyFromExpression(value, properties);
+      } else {
+        // otherwise continual loop;
+        getPropertyLoop(value, properties);
+      }
+    }
+  })
+}
+
+
+function getPropertyFromExpression(exp, properties) {
+  // now we care about the expression like:
+  // ["get", string] not ["get", string, Object],
+  // ["has", string] not ["has", string, Object],
+  // ["properties"],
+  // ["feature-state", string]
+  if (exp instanceof Array) {
+    switch (exp[0]) {
```

Comment:
Could `exp` ever be empty or zero length?

Reply:
Yes, good question, I should to judge the exp and I forget.
But while I review this line, I find it very interesting that, this time we are luck that we don't need to judge the exp, because:

1. if the exp is `null` or `undefined`, `exp instanceof Array` will return false.
2. if the exp is 0 length array, the `exp[0]` will be `undefined`, and in the next switch we only need to capture 4 conditions, it will not go into any condition, so we are fine.

Reply:
Ah, I see. So, sounds like this code is okay given the way that Javascript works. In C++ `exp[0]` would crash since C++ does not (for performance reasons) automatically bounds check arrays, so my PR review 👀 always look for that. 

Okay, so this is 👍 to keep the same.